### PR TITLE
IDSet: Update duplicate IDs check

### PR DESCRIPTION
**Pull request by yaakovi**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/46241

## Description
Consider the marketplace values when checking for duplicate IDs in the ID set.

### CODE REVIEW

File: demisto_sdk/commands/common/update_id_set.py [hunk 1]
```diff
@@ -133,8 +133,41 @@ def does_dict_have_alternative_key(data: dict) -> bool:
     return False
 
 
+def get_item_marketplaces(item_path: str, item_data: Dict = None, packs: Dict[str, Dict] = None):
```
Comment (reutshal) [inline demisto_sdk/commands/common/update_id_set.py:136]:
```suggestion
def get_item_marketplaces(item_path: str, item_data: Dict = None, packs: Dict[str, Dict] = None) -> List :
```

File: demisto_sdk/commands/common/update_id_set.py [hunk 2]
```diff
@@ -133,8 +133,41 @@ def does_dict_have_alternative_key(data: dict) -> bool:
...
+    else:
+        # default supporting marketplace
+        marketplaces = [MarketplaceVersions.XSOAR.value]
+
+    return marketplaces
+
+
 def should_skip_item_by_mp(file_path: str, marketplace: str, excluded_items_from_id_set: dict,
-                           print_logs: bool = False):
+                           packs: Dict[str, Dict] = None, print_logs: bool = False):
```
Comment (reutshal) [inline demisto_sdk/commands/common/update_id_set.py:170]: Please add to dosctring

File: demisto_sdk/commands/common/update_id_set.py [hunk 3]
```diff
@@ -389,7 +414,9 @@ def get_integration_data(file_path):
     integration_api_modules = get_integration_api_modules(file_path, data_dictionary, is_unified_integration)
     default_classifier = data_dictionary.get('defaultclassifier')
     default_incident_type = data_dictionary.get('defaultIncidentType')
+    is_fetch = data_dictionary.get('script', {}).get('isfetch', False)
```
Comment (reutshal) [inline demisto_sdk/commands/common/update_id_set.py:417]: Why did we add this?

File: demisto_sdk/commands/common/update_id_set.py [hunk 4]
```diff
@@ -1143,85 +1189,85 @@ def process_integration(file_path: str, print_logs: bool, marketplace: str = '')
...
-def process_incident_fields(file_path: str, print_logs: bool, incidents_types_list: list, marketplace: str = '') -> \
+def process_incident_fields(file_path: str, packs: Dict[str, Dict], marketplace: str, print_logs: bool, incident_types: List) -> \
         Tuple[list, dict]:
     """
     Process a incident_fields JSON file
     Args:
         file_path: The file path from incident field folder
         print_logs: Whether to print logs to stdout.
-        incidents_types_list: List of all the incident types in the system.
+        incident_types: List of all the incident types in the system.
```
Comment (reutshal) [inline demisto_sdk/commands/common/update_id_set.py:1250]: Add also 'packs' to docstring

File: demisto_sdk/commands/common/update_id_set.py [hunk 5]
```diff
@@ -1298,20 +1344,20 @@ def process_jobs(file_path: str, print_logs: bool, marketplace: str = '') -> lis
...
     result: List = []
     try:
-        if should_skip_item_by_mp(file_path, marketplace, {}, print_logs):
+        if should_skip_item_by_mp(file_path, marketplace, {}, packs=packs, print_logs=print_logs):
             return []
         if find_type(file_path) == FileType.JOB:
             if print_logs:
                 print(f'adding {file_path} to id_set')
-            result.append(get_job_data(file_path, print_logs))
+            result.append(get_job_data(file_path, packs=packs))
```
Comment (reutshal) [inline demisto_sdk/commands/common/update_id_set.py:1352]: Did we use to print twice?

File: demisto_sdk/commands/common/update_id_set.py [hunk 6]
```diff
@@ -1298,20 +1344,20 @@ def process_jobs(file_path: str, print_logs: bool, marketplace: str = '') -> lis
...
     except Exception as exp:  # noqa
         print_error(f'failed to process job {file_path}, Error: {str(exp)}')
         raise
     return result
 
 
-def process_general_items(file_path: str, print_logs: bool, expected_file_types: Tuple[FileType],
-                          data_extraction_func: Callable, marketplace: str = '') -> Tuple[list, dict]:
+def process_general_items(file_path: str, packs: Dict[str, Dict], marketplace: str, print_logs: bool,
+                          expected_file_types: Tuple[FileType], data_extraction_func: Callable) -> Tuple[list, dict]:
```
Comment (reutshal) [inline demisto_sdk/commands/common/update_id_set.py:1360]: Please update docstring accordingly

**Review by reutshal — CHANGES_REQUESTED** [review]
Add more tests

File: demisto_sdk/commands/common/update_id_set.py (hunk 3 above)
Comment (yaakovi) [reply demisto_sdk/commands/common/update_id_set.py:417]: It was requested several times in the last few days.
It's unrelated to the issue (I can move to another PR).

File: demisto_sdk/commands/common/update_id_set.py (hunk 5 above)
Comment (yaakovi) [reply demisto_sdk/commands/common/update_id_set.py:1352]: yes, it also appears in the calling function

File: demisto_sdk/commands/common/update_id_set.py [hunk 7]
```diff
@@ -133,8 +133,41 @@ def does_dict_have_alternative_key(data: dict) -> bool:
...
+    if not marketplaces and 'pack_metadata' not in item_path:
+        pack_name = get_pack_name(item_path)
+        if packs:
+            marketplaces = packs.get(pack_name, {}).get('marketplaces', [MarketplaceVersions.XSOAR.value])
+        else:
+            marketplaces = get_mp_types_from_metadata_by_item(item_path)
+    else:
+        # default supporting marketplace
+        marketplaces = [MarketplaceVersions.XSOAR.value]
+
```
Comment (ilappe) [inline demisto_sdk/commands/common/update_id_set.py:165]: @yaakovi 
if there is marketplaces defined - the if will go to the else and will return the default one

**Comment by ShahafBenYakir** [comment]
Can we merge this ASAP? private repo is failing and we have to merge some packs ASAP
@reutshal @yaakovi @ilappe

File: demisto_sdk/commands/common/update_id_set.py (hunk 7 above)
Comment (reutshal) [reply demisto_sdk/commands/common/update_id_set.py:165]: @ilappe  In what scenario do you see that happening?

Comment (yaakovi) [reply demisto_sdk/commands/common/update_id_set.py:165]: good catch! fixing now

**Review by ShahafBenYakir — APPROVED** [review]

File: demisto_sdk/commands/common/update_id_set.py [hunk 8]
```diff
@@ -133,8 +133,42 @@ def does_dict_have_alternative_key(data: dict) -> bool:
     return False
 
 
+def get_item_marketplaces(item_path: str, item_data: Dict = None, packs: Dict[str, Dict] = None) -> List:
+    """
+    Return the supporting marketplaces of the item.
+
+    Args:
+        item_path: the item path.
```
Comment (ilappe) [inline demisto_sdk/commands/common/update_id_set.py:141]: Can you move this function to `tools` as we discussed?

**Review by ilappe — APPROVED** [review]

File: demisto_sdk/commands/common/update_id_set.py (hunk 7 above)
Comment (ilappe) [reply demisto_sdk/commands/common/update_id_set.py]: @reutshal 
it was for example in case the item has ['marketplacev2'] in the marketplaces key

**Review by reutshal — APPROVED** [review]

**Pull request merged**